OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs

```diff
@@ -240,5 +244,26 @@ bool IsRallyPointValid(CPos x, BuildingInfo info)
 		{
 			return info != null && world.IsCellBuildable(x, null, info);
 		}
+
+		public bool HasAdequateRefineryCount
+		{
+			get
+			{
+				// Require at least one refinery, unless we can't build it.
+				return AIUtils.CountBuildingByCommonName(Info.RefineryTypes, player) >= MinimumRefineryCount ||
+					AIUtils.CountBuildingByCommonName(Info.PowerTypes, player) == 0 ||
+					AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) == 0;
+			}
+		}
+
+		int MinimumRefineryCount
+		{
+			get
+			{
+				// Unless we have no barracks (higher priority), require a 2nd refinery.
+				// TODO: Possibly unhardcode this, at least the targeted minimum of 2 (the fallback can probably stay at 1).
```

Comment:
General thought: IMO it would be better to define a list of build orders for main and expansion bases instead of random order + fraction / limit / priority bodges.